### PR TITLE
Only send HTTPDigestAuth on 4xx challenges

### PR DESCRIPTION
**Pull request by mmedal**
Resolves: #3772

### CODE REVIEW

**Comment by nateprewitt** [comment]
@Lukasa I believe there were issues with ONLY doing 401 as noted in #3772. The related RFCs allow for the WWW-Authenticate header in a non-401 response which the client SHOULD respond to if received. I think the conclusion we came to is to support 4XX response codes.

**Comment by mmedal** [comment]
I believe @nateprewitt is correct. It seemed to me that the conclusion of the discussion was to support 4XX response codes.

File: requests/auth.py [hunk 1]
```diff
@@ -227,6 +227,11 @@ def handle_401(self, r, **kwargs):
         :rtype: requests.Response
         """
 
+        # If response is not 4xx, do not auth
```
Comment (Lukasa) [inline requests/auth.py:230]: Given that I am clearly inclined to forget why we did this, let's add a link to the relevant issue here too. 😁

**Comment by mmedal** [comment]
Done!

**Comment by mmedal** [comment]
Might I also suggest renaming the hook to something like handle_4xx_auth_challenge?

**Comment by sigmavirus24** [comment]
> Might I also suggest renaming the hook to something like handle_4xx_auth_challenge?

I think I'd rather not. Most auth plugins use the same method name for consistency. But I'm not firmly set in that opinion.

**Comment by Lukasa** [comment]
Hrm. @mmedal, can you rebase on top of the current master? We seem to be having some build problems and I want to check whether they're ones we've seen before and fixed, or new ones.

**Comment by mmedal** [comment]
I triggered a new build (my branch is up to date), but it looks like you're still having build problems.

**Comment by nateprewitt** [comment]
@Lukasa this is a bug in pipenv (kennethreitz/pipenv#90) that may not have an immediate solution. It may be best to pin Requests' dependency for a bit. `pipenv==3.1.9` should work for now.

**Comment by Lukasa** [comment]
@nateprewitt I'll merge a PR that does that. 😁

**Comment by nateprewitt** [comment]
On it!

**Comment by mmedal** [comment]
Retriggering build.

**Pull request merged**